src/monodroid/jni/monodroid-glue-external.hh

```diff
@@ -0,0 +1,67 @@
+// Dear Emacs, this is a -*- C++ -*- header
+#ifndef __MONODROID_GLUE_EXTERNAL_HH
+#define __MONODROID_GLUE_EXTERNAL_HH
+
+#include <stdio.h>
+#include <stdarg.h>
+#include <jni.h>
+
+#include "monodroid.h"
+#include "logger.hh"
+#include "timing.hh"
+#include "xamarin_getifaddrs.h"
+
+namespace xamarin::android
+{
+	class MonodroidExternalAPI
```

Comment:
Naming quibble (?): the scenario being addressed is around P/Invokes to `__Internal` functions, in which mono eventually calls `MonodroidRuntime::monodroid_dlopen()`, which eventually calls `dlopen("libxa-mono-api.so")`.

This entire codepath is around P/Invokes to `__Internal`.  This is in no way "external".  (Unless we're using "external" as a "mirror" to "internal", at which point my brain potentially starts hurting.)

Then we have PR #4757, in which -- to address .NET mono "hijacking" `__Internal` for its own use, preventing us from using it -- changes our P/Invokes to use `monodroid` instead of `__Internal`.

Wouldn't this just "reintroduce" the same problem, as a `[DllImport("monodroid")]` would still result in a subsequent `monodroid_dlopen("monodroid")`, which would clear BSS, and break everything?

Furthermore, the whole reason we can't merge #4757 yet is *because* by using `[DllImport("monodroid")]`, we *break* desktop use, as there is no `monodroid.dll`/`libmonodroid.dylib` file for the Designer to load.

---

I thus suggest the following, in broad strokes:

 1. "Merge" this PR & PR #4757.
 2. As part of the merging, *instead of* `[DllImport("monodroid")]`, we do `[DllImport("xa-internal-api")]`, which in turn means we need to *build* `libxa-internal-api.dylib`/etc for desktop platforms.  (Or maybe `xa-private`, to "mirror" `System.Private.CoreLib`/etc.?)
 3. `libxa-mono-api.so` *is* `libxa-internal-api.so`.  
    (On further thought, I'm not really sure why `mono` is in the library name, when *most* of the member function are for XA use, not mono…)
 4. *This* type, `MonodroidExternalAPI`, should instead be `MonoAndroidInternalCalls` (kinda/sorta "mirroring" `Mono.Android.dll`).

Reply:
> Naming quibble (?): the scenario being addressed is around P/Invokes to `__Internal` functions, in which mono eventually calls `MonodroidRuntime::monodroid_dlopen()`, which eventually calls `dlopen("libxa-mono-api.so")`.
> 
> This entire codepath is around P/Invokes to `__Internal`. This is in no way "external". (Unless we're using "external" as a "mirror" to "internal", at which point my brain potentially starts hurting.)

The `__Internal` name has been a misnomer forever and it's irrelevant here. The class is named `External` because it exposes external API functions from XA/Monodroid - the fact they are used for "internal" p/invokes is not relevant in this light IMO.

> 
> Then we have PR #4757, in which -- to address .NET mono "hijacking" `__Internal` for its own use, preventing us from using it -- changes our P/Invokes to use `monodroid` instead of `__Internal`.
> 
> Wouldn't this just "reintroduce" the same problem, as a `[DllImport("monodroid")]` would still result in a subsequent `monodroid_dlopen("monodroid")`, which would clear BSS, and break everything?

It would, and 4757 needs to be updated, obviously. This issue herewasn't known when #4757 was opened...

> 
> Furthermore, the whole reason we can't merge #4757 yet is _because_ by using `[DllImport("monodroid")]`, we _break_ desktop use, as there is no `monodroid.dll`/`libmonodroid.dylib` file for the Designer to load.

While this PR builds `libxa-mono-api.{dll,dylib}`... So we should be able to update #4757 to use `libxa-mono-api` instead of `monodroid`

> 
> I thus suggest the following, in broad strokes:
> 
>     1. "Merge" this PR & PR #4759 
We can do that, but this PR fixes a real user issue while #4757 is for the world 12 months from now really. I'm fine with it if you think it's ok to merge them, though

> 
>     2. As part of the merging, _instead of_ `[DllImport("monodroid")]`, we do `[DllImport("xa-internal-api")]`, which in turn means we need to _build_ `libxa-internal-api.dylib`/etc for desktop platforms.  (Or maybe `xa-private`, to "mirror" `System.Private.CoreLib`/etc.?)
Already builds for desktop

> 
>     3. `libxa-mono-api.so` _is_ `libxa-internal-api.so`.
>        (On further thought, I'm not really sure why `mono` is in the library name, when _most_ of the member function are for XA use, not mono…)
It's not internal API as far as XA is concerned - it's an export of API calls, so it's by all means external - to be used by external code. However, I don't really care enough about the name, I'll just update it to internal in a while.

> 
>     4. _This_ type, `MonodroidExternalAPI`, should instead be `MonoAndroidInternalCalls` (kinda/sorta "mirroring" `Mono.Android.dll`).
:+1:

Reply:
> "Merge" this PR & PR #4759 

In retrospect, let's not do this.  Instead, let's keep it in mind: the implicit intent *here* is that we shouldn't necessarily be P/Invoking to `monodroid` or `__Internal`, but instead we should *eventually* be P/Invoking to… (1) *something* that we control, and (2) a "real" filename, not a "virtual" one, which is the problem with `monodroid` & PR #4759: there is no "real" `monodroid` library *unless* you're on-device, which isn't the case for the Designer.

In this case, `libxa-internal-api.so` should be our preferred "stable" P/Invoke target.